controller/api/destination/endpoint_translator.go

```diff
@@ -79,33 +110,120 @@ func newEndpointTranslator(
+	et.enqueueUpdate(&noEndpointsUpdate{exists})
+}
+
+// Add, Remove, and NoEndpoints are called from a client-go informer callback
+// and therefore must not block. For each of these, we enqueue an update in
+// a channel so that it can be processed asyncronously. To ensure that enqueuing
+// does not block, we first check to see if there is capacity in the buffered
+// channel. If there is not, we drop the update and signal to the stream that
+// it has fallen too far behind and should be closed.
+func (et *endpointTranslator) enqueueUpdate(update interface{}) {
+	select {
+	case et.updates <- update:
+		// Update has been successfully enqueued.
+	default:
+		// We are unable to enqueue because the channel does not have capacity.
+		// The stream has fallen too far behind and should be closed.
+		et.overflowCounter.Inc()
+		select {
+		case <-et.endStream:
+			// The endStream channel has already been closed so no action is
+			// necessary.
+		default:
+			et.log.Error("endpoint update queue full; ending stream")
+			close(et.endStream)
+		}
+	}
+}
+
+// Start initiates a goroutine which processes update events off of the
+// endpointTranslator's internal queue and sends to the grpc stream as
+// appropriate. The goroutine calls several non-thread-safe functions (including
+// Send) and therefore, Start must not be called more than once.
+func (et *endpointTranslator) Start() {
+	go func() {
+		for {
+			select {
+			case update := <-et.updates:
+				et.processUpdate(update)
+			case <-et.stop:
+				return
+			}
+		}
+	}()
+}
 
+// Stop terminates the goroutine started by Start.
+func (et *endpointTranslator) Stop() {
+	close(et.stop)
+}
+
+func (et *endpointTranslator) processUpdate(update interface{}) {
+	switch update := update.(type) {
+	case *addUpdate:
+		et.add(update.set)
+	case *removeUpdate:
+		et.remove(update.set)
+	case *noEndpointsUpdate:
+		et.noEndpoints(update.exists)
+	}
+}
+
+func (et *endpointTranslator) add(set watcher.AddressSet) {
 	for id, address := range set.Addresses {
 		et.availableEndpoints.Addresses[id] = address
 	}
 
 	et.sendFilteredUpdate(set)
 }
 
-func (et *endpointTranslator) Remove(set watcher.AddressSet) {
-	et.mu.Lock()
-	defer et.mu.Unlock()
-
+func (et *endpointTranslator) remove(set watcher.AddressSet) {
 	for id := range set.Addresses {
 		delete(et.availableEndpoints.Addresses, id)
 	}
 
 	et.sendFilteredUpdate(set)
 }
 
+func (et *endpointTranslator) noEndpoints(exists bool) {
+	et.log.Debugf("NoEndpoints(%+v)", exists)
+
+	et.availableEndpoints.Addresses = map[watcher.ID]watcher.Address{}
+	et.filteredSnapshot.Addresses = map[watcher.ID]watcher.Address{}
+
+	u := &pb.Update{
+		Update: &pb.Update_NoEndpoints{
+			NoEndpoints: &pb.NoEndpoints{
+				Exists: exists,
+			},
+		},
+	}
+
+	et.log.Debugf("Sending destination no endpoints: %+v", u)
+	if err := et.stream.Send(u); err != nil {
+		et.log.Debugf("Failed to send address update: %s", err)
```

Comment:
I know this is previous code that we've moved, but why would an error would be logged at Debug level?

Reply:
We expect to see this every time a client proxy restarts/closes a stream.